### PR TITLE
Use dataclasses for all objects returned by HfApi #1911

### PR DESCRIPTION
**Pull request by Ahmedniz1**

### CODE REVIEW

**Comment by Ahmedniz1** [comment]
@Wauplin I've updated the code for bloblfsinfo class. What do I need to make sure to commit the PR. 
Thanks

File: src/huggingface_hub/hf_api.py [hunk 1]
```diff
@@ -243,12 +243,16 @@ class LastCommitInfo(TypedDict, total=False):
     title: str
     date: datetime
 
-
-class BlobLfsInfo(TypedDict, total=False):
+@dataclass
+class BlobLfsInfo(Dict):
```
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:247]:
```suggestion
class BlobLfsInfo(dict):
```

File: src/huggingface_hub/hf_api.py [hunk 2]
```diff
@@ -243,12 +243,16 @@ class LastCommitInfo(TypedDict, total=False):
...
     size: int
     sha256: str
     pointer_size: int
-
+    def __init__(self, data:dict):
```
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:251]:
```suggestion
    def __init__(self, data:dict):  # hack to make BlobLfsInfo backward compatible
```

**Review by Wauplin — COMMENTED** [review]
Hey @Ahmedniz1 thanks for starting on this PR. I've left 2 small comments on it. Apart from that, you'll need to:
1. Make sure all instantiations of `BlobLfsInfo` are now passing a dictionary as input. For example, [this line](https://github.com/huggingface/huggingface_hub/blob/f76dc44efb9310ec222ad17f6f87e2964b443ac8/src/huggingface_hub/hf_api.py#L501) must be replaced by `lfs = BlobLfsInfo(lfs)` 
2. Format your code to comply with the codebase rules. To do so, install the dev dependencies (`pip install -e ".[dev]"`) and run the formatter (`make style`). Then you can check everything's good (`make quality`) and finally commit the changes done to the code (most likely additional whitespace + newlines).
3. Finally it is generally a good practice to add a short description to the PR ([here](https://github.com/huggingface/huggingface_hub/pull/1974#issue-2079461566)) and especially mention that it's solving issue https://github.com/huggingface/huggingface_hub/issues/1911. This way we have a connection between the issue and the related PR.

That should be it! Thanks for your contribution! :)

File: src/huggingface_hub/hf_api.py [hunk 3]
```diff
@@ -244,11 +244,18 @@ class LastCommitInfo(TypedDict, total=False):
...
     size: int
     sha256: str
     pointer_size: int
 
+    def __init__(self, data: dict):  # hack to make BlobLfsInfo backward compatible
+        self.__dict__.update(data)
+        self.size = data["size"]
+        self.sha256 = data["sha256"]
+        self.pointer_size = data["pointer_size"]
+
```
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:258]: I realized changing to `def __init__(self, data: dict):` is really backward compatible. What we can do instead is to use the `__post_init__` method to update the internal dict once all attributes have been set. It is also cleaner this way as we don't have to manually redefine the attributes.

```suggestion
    def __post_init__(self):
        # hack to make BlobLfsInfo backward compatible
        self.update(asdict(self))

```

:warning: you must add `from dataclasses import asdict` at the top of the file

File: src/huggingface_hub/hf_api.py [hunk 4]
```diff
@@ -498,7 +505,7 @@ def __init__(self, **kwargs):
         self.blob_id = kwargs.pop("oid")
         lfs = kwargs.pop("lfs", None)
         if lfs is not None:
-            lfs = BlobLfsInfo(size=lfs["size"], sha256=lfs["oid"], pointer_size=lfs["pointerSize"])
+            lfs = BlobLfsInfo(lfs)
```
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:508]: Related to the comment above, this modification is not needed anymore.

```suggestion
            lfs = BlobLfsInfo(size=lfs["size"], sha256=lfs["oid"], pointer_size=lfs["pointerSize"])
```

File: src/huggingface_hub/hf_api.py [hunk 5]
```diff
@@ -670,9 +677,11 @@ def __init__(self, **kwargs):
...
                     lfs=(
                         BlobLfsInfo(
-                            size=sibling["lfs"]["size"],
-                            sha256=sibling["lfs"]["sha256"],
-                            pointer_size=sibling["lfs"]["pointerSize"],
+                            {
+                                "size": sibling["lfs"]["size"],
+                                "sha256": sibling["lfs"]["sha256"],
+                                "pointer_size": sibling["lfs"]["pointerSize"],
+                            }
```
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:684]: Related to the comment above, this modification is not needed anymore.

```suggestion
                            size=sibling["lfs"]["size"],
                            sha256=sibling["lfs"]["sha256"],
                            pointer_size=sibling["lfs"]["pointerSize"],
```

File: src/huggingface_hub/hf_api.py [hunk 6]
```diff
@@ -781,9 +790,11 @@ def __init__(self, **kwargs):
...
```
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:797]: Related to the comment above, this modification is not needed anymore.

```suggestion
                            size=sibling["lfs"]["size"],
                            sha256=sibling["lfs"]["sha256"],
                            pointer_size=sibling["lfs"]["pointerSize"],
```

File: src/huggingface_hub/hf_api.py [hunk 7]
```diff
@@ -903,9 +914,11 @@ def __init__(self, **kwargs):
...
```
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:921]: Related to the comment above, this modification is not needed anymore.

```suggestion
                            size=sibling["lfs"]["size"],
                            sha256=sibling["lfs"]["sha256"],
                            pointer_size=sibling["lfs"]["pointerSize"],
```

**Review by Wauplin — COMMENTED** [review]
Hi @Ahmedniz1, thanks for making the changes! The CI failed tests made me realize that the current solution that I've suggested to make is not satisfying in term of backward compatibility. I'm suggesting you some new changes that should make it cleaner and more robust for downstream apps. I'm sorry about this change of instructions, the fault is on my end. Better to fix it now rather than when it's fixed though :)

File: src/huggingface_hub/hf_api.py (hunk 7 above)
Comment (Ahmedniz1) [reply src/huggingface_hub/hf_api.py]: Thanks for the review, I've updated it accordingly.

File: src/huggingface_hub/hf_api.py [hunk 8]
```diff
@@ -244,11 +244,21 @@ class LastCommitInfo(TypedDict, total=False):
...
     pointer_size: int
 
+    def __init__(self, data: dict):  # hack to make BlobLfsInfo backward compatible
+        self.__dict__.update(data)
+        self.size = data["size"]
+        self.sha256 = data["sha256"]
+        self.pointer_size = data["pointer_size"]
+
+    def __post_init__(self):
+        self.update(asdict(self))
```
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:260]: (no need of the `__init__` anymore)

```suggestion
    def __post_init__(self):  # hack to make BlobLfsInfo backward compatible
        self.update(asdict(self))
```

**Review by Wauplin — APPROVED** [review]
Thanks for making the changes @Ahmedniz1! Ending up with a small and clean PR in the end which is nice :) Let's wait for the CI to complete and then we should be fine to merge this! :tada:

**Comment by Wauplin** [comment]
It's green! Merging it now. Thanks again :)

**Comment by Ahmedniz1** [comment]
Thanks for guiding me in detail @Wauplin. I really appreciate that. 
Now, I just want to confirm this, should I update other classes to use data class which were mentioned in issue #1911.
Thanks

**Comment by Wauplin** [comment]
@Ahmedniz1 yes you can! Thanks for the help. I think NouamaneELGueddarii is working on the `TransformerInfo` class but you can work on the other if you'd like. You can make several classes in the same PR now that the process is clearer (both for you and for me). Just make sure to post on #1911 about it so that everyone knows on what you are working. Thanks in advance!

**Pull request merged**